src/openms/source/FORMAT/MzQCFile.cpp

```diff
@@ -153,92 +101,140 @@ namespace OpenMS
         } 
         else
         {
-          qm["name"] = name;
           cout << accession << " not found in CV." << endl;
+          return;
         }
         qm["value"] = value;
-        return qm;
+        quality_metrics.push_back(qm);
       };
 
-    map<Size, UInt> counts;
-    for (const auto& spectrum : exp)
-      {
-        const Size level = spectrum.getMSLevel();
-        ++counts[level];  // count MS level
-      }
-    // QC:4000059 Number of MS1 spectra
-    out["mzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000059", "Number of MS1 spectra", counts[1]);
-    // QC:4000060 Number of MS2 spectra
-    out["mzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000060", "Number of MS2 spectra", counts[2]);
-    // QC:4000135 Number of chromatograms"
-    out["mzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000135", "Number of chromatograms", exp.getChromatograms().size());
-    // QC:4000053 Run time (RT duration)
-    out["mzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000053", "RT duration", UInt(exp.getMaxRT() - exp.getMinRT()));
-    // QC:4000138 MZ acquisition range
-    out["mzQC"]["runQualities"]["qualityMetrics"] += addValue("QC:4000138", "MZ acquisition range", tuple<UInt,UInt>{exp.getMinMZ(), exp.getMaxMZ()});
-    
-    const MSChromatogram& tic = exp.getTIC();
-    UInt jump = 0;
-    UInt fall = 0;
-    vector<float> retention_times;
-    vector<UInt> intensities;
-    
-    for (const auto& p : tic)
+    // ---------------------------------------------------------------
+    // collecting quality metrics
+    // ---------------------------------------------------------------
+
+    if (spectrum_count.isRunnable(status))
+    {
+      auto counts = spectrum_count.compute(exp);
+      // Number of MS1 spectra
+      addMetric("QC:4000059", counts[1]);
+      // Number of MS2 spectra
+      addMetric("QC:4000060", counts[2]);
+    }
+    // Number of chromatograms"
+    addMetric("QC:4000135", exp.getChromatograms().size());
+    // Run time (RT duration)
+    addMetric("QC:4000053", UInt(exp.getMaxRT() - exp.getMinRT()));
+    // MZ acquisition range
+    addMetric("QC:4000138", tuple<UInt,UInt>{exp.getMinMZ(), exp.getMaxMZ()});
+
+    if (tic.isRunnable(status) and !exp.getTIC().empty())
```

Comment:
`exp.getTIC()` is essentially computed twice, right?
Why not just call `tic.compute()` and see if result is empty?
Or, change the `compute()` to accept an externally computed TIC...

Reply:
true, it is computed twice, removed that and checked if the result contains a vector of intensities.